### PR TITLE
1191 arv medication

### PR DESCRIPTION
**Pull request by clemens-msupply**
Part of  #1191

Includes:
- point in the issue except of the Pill Counts section which will be done in another PR
- a ProgramEvent api to query program events for an encounter
- an EncounterProgramEvent component to display the event data at a given point in time
- Use program events to display "Current ARV Regime" and "Pill Count after last Visit". This makes it possible to have "gaps" in the encounter where none of this data is recorded. For example, if a previous encounter didn't change the ARV regime the next encounter still displays the correct current ARV Regime.
- refactor of common JSON form layout constants

In the issue is one point I haven't worked on: "If Regimen Status = Start ARV or Changes to ARV Regimen, it has to be different from Current ARV Regimen (data validation)" This might be tricky and not sure if it is really required or a nice to have @richardmoizeau ?

### CODE REVIEW

File: client/packages/programs/src/JsonForms/common/styleConstants.ts [hunk 1]
```diff
@@ -1,3 +1,13 @@
+import { Theme } from '@mui/material';
+import { SxProps } from '@mui/system';
+
 export const FORM_LABEL_WIDTH = 40;
 export const FORM_LABEL_COLUMN_WIDTH = `${FORM_LABEL_WIDTH}%`;
 export const FORM_INPUT_COLUMN_WIDTH = `${100 - FORM_LABEL_WIDTH}%`;
+
+/** Default sx style for a single form row. */
+export const DefaultFormRowSx: SxProps<Theme> = {
```
Comment (clemens-msupply) [inline client/packages/programs/src/JsonForms/common/styleConstants.ts:9]: Was not sure about the naming convention for this const...

**Comment by mark-prins** [comment]
In the issue we have 

>  End of Supply: calculated based on Pill Count (at visit end) except if

Do you know what the `if` is?

**Comment by clemens-msupply** [comment]
> In the issue we have
> 
> > End of Supply: calculated based on Pill Count (at visit end) except if
> 
> Do you know what the `if` is?

No, I assumed the "except if" was some unintended extra text? but maybe I miss something? @richardmoizeau

File: client/packages/programs/src/JsonForms/common/styleConstants.ts [hunk 2]
```diff
@@ -1,3 +1,13 @@
...
+  margin: 0.5,
```
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/common/styleConstants.ts:10]: For almost all of the usages you could add

```
  alignItems: 'center',
  display: 'flex',
  justifyContent: 'space-around',
  minWidth: '300px',
```

afaics it is only the `JSONForms/Text.tsx` and the `inputProps` on `DetailInputWithLabelRow` in `EncounterProgramEvent` which don't have these additional props. 

In my testing so far, this looks fine - and it means that you can remove a lot more of shared style items:

```
export const DefaultFormRowSpacing: SxProps<Theme> = {
  gap: 2,
  margin: 0.5,
  marginLeft: 0,
};

/** Default sx style for a single form row. */
export const DefaultFormRowSx: SxProps<Theme> = {
  alignItems: 'center',
  display: 'flex',
  justifyContent: 'space-around',
  minWidth: '300px',
  ...DefaultFormRowSpacing,
};
```


with the `EncounterProgramEvent` being

```
return (
    <DetailInputWithLabelRow
      label={label}
      inputProps={{
        value: event?.data ?? '',
        disabled: true,
        required: props.required,
        sx: DefaultFormRowSpacing,
        error: !!errors,
        helperText: errors,
      }}
      labelWidthPercentage={FORM_LABEL_WIDTH}
      inputAlignment={'start'}
    />
  );
```

File: client/packages/programs/src/JsonForms/components/PreviousEncounterField.tsx [hunk 3]
```diff
@@ -53,7 +54,7 @@ const UIComponent = (props: ControlProps) => {
       label={label}
       inputProps={{
         ...inputProps,
-        sx: { ...inputProps.sx, margin: 0.5, width: '100%' },
+        sx: { ...inputProps.sx, margin: 0.5, marginLeft: 0, gap: 2 },
```
Comment (mark-prins) [inline client/packages/programs/src/JsonForms/components/PreviousEncounterField.tsx:57]:
```suggestion
        sx: { ...inputProps.sx, ...DefaultFormRowSx },
```

**Review by mark-prins — COMMENTED** [review]
Looks great - covers everything I can see, and I like the introduction of the styleConstants.
One thing - I couldn't see `Number of pills dispensed` am I missing something?

**Review by mark-prins — APPROVED** [review]
👍 ah, ok. I see the quantity dispensed is in PR2. didn't realise that after reading the description on this PR

**Pull request merged**